You-Can-Code-It/fndr
# Company tags

# What this pull request does

https://github.com/luizfiorentino/fndr/assets/96445830/ca4e5a4b-1c60-4b15-99ae-8e21feebd0b9

## What

- adds tags feature for the companies
- related tags can be added in the EditCompanyForm with a "title" and a "category"
- it is also possible to remove a tag in the same form
- the tags can be displayed in the Card component (displayed in the index page), in the Details Page and in the EditCompanyForm

## Why

- by adding tags to a company, users may get useful information of a company, for example the programming or spoken languages, which helps when deciding whether to apply for a vacancy or not
- by removing a tag from a company's profile, users get an updated overview of the company's infos

## How

(a) database
- a new table "Tag" is added with "title" (with "unique" constraint to avoid unnecessary duplicates) and "category"
- the "Tag" and "Company" tables have a "many-to-many" relationship

(b) api endpoints
`/api/companies/companyId`: adds or relate a tag to a company. It uses the "connectOrCreate" statement so that an existing tag can be related to the company
`/api/companies/companyId/tags`: disconnects a tag from a company's data

(c) frontend
- in the edit company form a button is created to add a new tag. When it's clicked, two input fields are shown: "title" and "category"
- to add a tag: uses the existing `formSubmit` function that makes a patch request to the api route with the form data
- to disconnect an existing tag from the company's data: a new "removeTag" function is created to make an update request to the api. It passes the company's id using the component's props (company.id) and the tags' id

## Steppings to test this

- update your database schema with the commands `npx prisma migrate dev`, `npx prisma generate` and `npx prisma studio` to check the new "Tag" table is created and the many-to-many relationship between it and the "Company" table is established
- select a company to edit its data: a new field is added to the form. Click `Add tag` and fill in title and category (e.g. "React" and "Programming Language" respectively)
- submit the form. Now the tag is added to the form, the details page and to the company card in the index page
- in the Edit Company Form page, there is a trash bin icon. Click on it and the tag will be removed

## Anything else

- to avoid similar tags with different names, for example "Programming language" and "Programming skills" pointing to "JavaScript", and similar ambiguities when adding tags, it would be interesting to figure out a way of standardising their creation, for example by replacing the "category" field in the EditCompanyForm, so that the user has a limited rol of choices
- before removing the tag the user can receive a message to confirm it
- the feature still needs styling
